diff -urNp openswan-2.6.32-patched/lib/libpluto/packet.c openswan-2.6.32-current/lib/libpluto/packet.c
--- openswan-2.6.32-patched/lib/libpluto/packet.c	2012-06-22 11:13:26.612890460 -0400
+++ openswan-2.6.32-current/lib/libpluto/packet.c	2012-06-22 11:17:13.657819028 -0400
@@ -662,7 +662,7 @@ struct_desc ikev2_sa_desc = { "IKEv2 Sec
  */
 static field_desc ikev2prop_fields[] = {
     { ft_enum, 8/BITS_PER_BYTE, "next payload type", &payload_names },
-    { ft_mbz,  8/BITS_PER_BYTE, NULL, NULL },
+    { ft_nat,  8/BITS_PER_BYTE, "reserved", NULL },
     { ft_len, 16/BITS_PER_BYTE, "length", NULL },
     { ft_nat,  8/BITS_PER_BYTE, "prop #", NULL },
     { ft_nat,  8/BITS_PER_BYTE, "proto ID", NULL },
@@ -693,10 +693,10 @@ struct_desc ikev2_prop_desc = { "IKEv2 P
  */
 static field_desc ikev2trans_fields[] = {
     { ft_enum, 8/BITS_PER_BYTE, "next payload type", &payload_names },
-    { ft_mbz,  8/BITS_PER_BYTE, NULL, NULL },
+    { ft_nat,  8/BITS_PER_BYTE, "reserved", NULL },
     { ft_len, 16/BITS_PER_BYTE, "length", NULL },
     { ft_nat,  8/BITS_PER_BYTE, "transform type", &trans_type_names },
-    { ft_mbz,  8/BITS_PER_BYTE, NULL, NULL },
+    { ft_nat,  8/BITS_PER_BYTE, "reserved", NULL },
     { ft_nat, 16/BITS_PER_BYTE, "transform ID", NULL },
     { ft_end,  0, NULL, NULL }
 };
@@ -752,11 +752,10 @@ struct_desc ikev2_trans_attr_desc = {
  */
 static field_desc ikev2ke_fields[] = {
     { ft_enum, 8/BITS_PER_BYTE, "next payload type", &payload_names },
-    //{ ft_mbz,  8/BITS_PER_BYTE, NULL, NULL },
     { ft_set, 8/BITS_PER_BYTE, "critical bit", critical_names},
     { ft_len, 16/BITS_PER_BYTE, "length", NULL },
     { ft_nat, 16/BITS_PER_BYTE, "transform type", &oakley_group_names },
-    { ft_mbz, 16/BITS_PER_BYTE, NULL, NULL },
+    { ft_nat, 16/BITS_PER_BYTE, "reserved", NULL },
     { ft_end,  0, NULL, NULL }
 };
 
@@ -796,12 +795,11 @@ struct_desc ikev2_ke_desc = { "IKEv2 Key
 
 static field_desc ikev2id_fields[] = {
     { ft_enum, 8/BITS_PER_BYTE, "next payload type", &payload_names },
-    //{ ft_mbz,  8/BITS_PER_BYTE, NULL, NULL },
     { ft_set, 8/BITS_PER_BYTE, "critical bit", critical_names},
     { ft_len, 16/BITS_PER_BYTE, "length", NULL },
     { ft_enum, 8/BITS_PER_BYTE, "id_type", &ident_names },
-    { ft_mbz,  8/BITS_PER_BYTE, NULL, NULL },
-    { ft_mbz, 16/BITS_PER_BYTE, NULL, NULL },
+    { ft_nat,  8/BITS_PER_BYTE, "reserved", NULL },
+    { ft_nat, 16/BITS_PER_BYTE, "reserved", NULL },
     { ft_end,  0, NULL, NULL }
 };
 
@@ -882,12 +880,11 @@ struct_desc ikev2_certificate_req_desc =
  */
 static field_desc ikev2a_fields[] = {
     { ft_enum, 8/BITS_PER_BYTE, "next payload type", &payload_names },
-    //{ ft_mbz,  8/BITS_PER_BYTE, NULL, NULL },
     { ft_set, 8/BITS_PER_BYTE, "critical bit", critical_names},
     { ft_len, 16/BITS_PER_BYTE, "length", NULL },
     { ft_enum, 8/BITS_PER_BYTE, "auth method", &ikev2_auth_names },
-    { ft_mbz,  8/BITS_PER_BYTE, NULL, NULL },
-    { ft_mbz, 16/BITS_PER_BYTE, NULL, NULL },
+    { ft_nat,  8/BITS_PER_BYTE, "reserved", NULL },
+    { ft_nat, 16/BITS_PER_BYTE, "reserved", NULL },
     { ft_end,  0, NULL, NULL }
 };
 
@@ -1030,12 +1027,11 @@ struct_desc ikev2_vendor_id_desc = { "IK
  */
 static field_desc ikev2ts_fields[] = {
     { ft_enum, 8/BITS_PER_BYTE, "next payload type", &payload_names },
-    //{ ft_mbz,  8/BITS_PER_BYTE, NULL, NULL },
     { ft_set, 8/BITS_PER_BYTE, "critical bit", critical_names},
     { ft_len, 16/BITS_PER_BYTE, "length", NULL },
     { ft_nat,  8/BITS_PER_BYTE, "number of TS", NULL},
-    { ft_mbz,  8/BITS_PER_BYTE, NULL, NULL },
-    { ft_mbz, 16/BITS_PER_BYTE, NULL, NULL },
+    { ft_nat,  8/BITS_PER_BYTE, "reserved", NULL },
+    { ft_nat, 16/BITS_PER_BYTE, "reserved", NULL },
     { ft_end,  0, NULL, NULL }
 };
 struct_desc ikev2_ts_desc = { "IKEv2 Traffic Selectors",
