diff -Naur openswan-2.6.32-orig/programs/pluto/ikev2.c openswan-2.6.32/programs/pluto/ikev2.c
--- openswan-2.6.32-orig/programs/pluto/ikev2.c	2014-06-20 14:35:23.924825774 -0400
+++ openswan-2.6.32/programs/pluto/ikev2.c	2014-06-20 16:13:10.850934298 -0400
@@ -852,8 +852,9 @@
 	/* XXX should call unpend again on parent SA */
 	if(st->st_clonedfrom != 0) {
 	    pst = state_with_serialno(st->st_clonedfrom);
-	    DBG_log("releasing whack for #%lu (sock=%d)"
+	    DBG_log("releasing whack and unpending for #%lu (sock=%d)"
 		    , pst->st_serialno, pst->st_whack_sock);
+	    unpend(pst);
 	    release_whack(pst);
 	}
     }
diff -Naur openswan-2.6.32-orig/programs/pluto/ipsec_doi.c openswan-2.6.32/programs/pluto/ipsec_doi.c
--- openswan-2.6.32-orig/programs/pluto/ipsec_doi.c	2014-06-20 14:35:24.016827904 -0400
+++ openswan-2.6.32/programs/pluto/ipsec_doi.c	2014-06-20 16:16:10.831729625 -0400
@@ -306,7 +306,8 @@
 static initiator_function *pick_initiator(struct connection *c UNUSED, lset_t policy)
 {
     if((policy & POLICY_IKEV1_DISABLE) == 0 &&
-       (c->failed_ikev2 || (policy & POLICY_IKEV2_PROPOSE)==0))  {
+       (c->failed_ikev2 || (policy & POLICY_IKEV2_PROPOSE) == 0) &&
+       (c->policy & (POLICY_IKEV1_DISABLE | POLICY_IKEV2_PROPOSE)) == 0) {
 	if(policy & POLICY_AGGRESSIVE) {
 #if defined(AGGRESSIVE)	    
 	    return aggr_outI1;
@@ -317,7 +318,8 @@
 	    return main_outI1;
 	}
 
-    } else if(policy & POLICY_IKEV2_PROPOSE) {
+    } else if ((policy & POLICY_IKEV2_PROPOSE) ||
+	       (c->policy & (POLICY_IKEV1_DISABLE | POLICY_IKEV2_PROPOSE))) {
 	return ikev2parent_outI1;
 
     } else {
@@ -461,6 +463,10 @@
 	    if (st->st_ipcomp.attrs.encapsulation == ENCAPSULATION_MODE_TUNNEL)
 		policy |= POLICY_TUNNEL;
 	}
+	/* retain policy so child sa rekey attempt does not blow up */
+	if (st->st_state == STATE_PARENT_I3)
+		policy = st->st_connection->policy;
+
 	passert(HAS_IPSEC_POLICY(policy));
 	ipsecdoi_initiate(whack_sock, st->st_connection, policy, try
 			  , st->st_serialno, st->st_import 
diff -Naur openswan-2.6.32-orig/programs/pluto/pending.c openswan-2.6.32/programs/pluto/pending.c
--- openswan-2.6.32-orig/programs/pluto/pending.c	2014-06-20 14:35:23.406813785 -0400
+++ openswan-2.6.32/programs/pluto/pending.c	2014-06-20 16:17:49.096753342 -0400
@@ -207,18 +207,21 @@
 	if (p->isakmp_sa == st)
 	{
 	    DBG(DBG_CONTROL
-		, DBG_log("unqueuing pending Quick Mode with %s \"%s\" %s"
+		, DBG_log("unqueuing pending %s with %s \"%s\" %s"
+			  , st->st_ikev2 ? "Child SA" : "Quick Mode"
 			  , ip_str(&p->connection->spd.that.host_addr)
 			  , p->connection->name
 			  , enum_name(&pluto_cryptoimportance_names,st->st_import)));
 
 	    p->pend_time = time(NULL);
-	    (void) quick_outI1(p->whack_sock, st, p->connection, p->policy
+	    if (!st->st_ikev2) {
+	    	(void) quick_outI1(p->whack_sock, st, p->connection, p->policy
 			       , p->try, p->replacing
 #ifdef HAVE_LABELED_IPSEC
 			       , p->uctx
 #endif
 			      );
+	    }
 	    p->whack_sock = NULL_FD;	/* ownership transferred */
 	    p->connection = NULL;	/* ownership transferred */
 	    delete_pending(pp);
