diff -Naur openswan-2.6.32-orig/programs/_updown.netkey/_updown.netkey.in openswan-2.6.32/programs/_updown.netkey/_updown.netkey.in
--- openswan-2.6.32-orig/programs/_updown.netkey/_updown.netkey.in	2014-04-18 12:06:42.502588290 -0400
+++ openswan-2.6.32/programs/_updown.netkey/_updown.netkey.in	2014-04-18 12:07:38.345852273 -0400
@@ -180,7 +180,7 @@
 }
 
 downrule() {
-        if [ -n "$PLUTO_MY_SOURCEIP" ]
+        if [ -n "$PLUTO_MY_SOURCEIP" -a "$PLUTO_IS_PEER_CISCO" -eq "1" ]
         then
             doroute del
             ip route flush cache
